rproxy/raw_proxy_test.go

```diff
@@ -74,7 +74,14 @@ func TestRawProxy_Terminate(t *testing.T) {
 		}
 		time.Sleep(20 * time.Millisecond)
 	}
-	assert.Equal(t, proxy.GetInfo().RawConnections, 0)
+
+	deadline = time.Now().Add(time.Second)
+	for proxy.GetInfo().RawConnections > 0 {
+		if time.Now().After(deadline) {
+			t.Fatal("Expected RawConnections to drop to 0")
+		}
+		time.Sleep(20 * time.Millisecond)
+	}
```

Comment:
So the story is there's a time window when `conn.Call(resp.MsgFromStrings("ping")` throws errors but proxy.GetInfo() still thinks there's a connection, aye?

The change makes sense, but just as an exercise I tried to reproduce this locally. I wasn't able to do this the dumb way (run the original test in a loop, decrease or remove the sleep). A smarter way to reproduce the race could be to add a small sleep somewhere else in order to make this time window more explicit, right?

Reply:
Yup.  Like this:
```diff
diff --git a/rproxy/raw_handler.go b/rproxy/raw_handler.go
index f471687..c1579de 100644
--- a/rproxy/raw_handler.go
+++ b/rproxy/raw_handler.go
@@ -4,6 +4,7 @@ import (
        "io"
        "log"
        "net"
+       "time"
 )

 type RawHandler struct {
@@ -34,6 +35,7 @@ func (r *RawHandler) Run() {
        defer func() {
                r.cliConn.Close()
                r.uplinkConn.Close()
+               time.Sleep(time.Second)
                r.proxy.rawProxy.deadHandlerChan <- r
        }()

diff --git a/rproxy/raw_proxy_test.go b/rproxy/raw_proxy_test.go
index 9c9699e..865fe9f 100644
--- a/rproxy/raw_proxy_test.go
+++ b/rproxy/raw_proxy_test.go
@@ -75,14 +75,17 @@ func TestRawProxy_Terminate(t *testing.T) {
                time.Sleep(20 * time.Millisecond)
        }

-       deadline = time.Now().Add(time.Second)
-       for proxy.GetInfo().RawConnections > 0 {
-               if time.Now().After(deadline) {
-                       t.Fatal("Expected RawConnections to drop to 0")
-               }
-               time.Sleep(20 * time.Millisecond)
-       }
-
+       assert.Equal(t, proxy.GetInfo().RawConnections, 0)
+       //
+       //
+       //      deadline = time.Now().Add(time.Second)
+       //      for proxy.GetInfo().RawConnections > 0 {
+       //              if time.Now().After(deadline) {
+       //                      t.Fatal("Expected RawConnections to drop to 0")
+       //              }
+       //              time.Sleep(20 * time.Millisecond)
+       //      }
+       //
        proxy.Stop()
        waitUntil(t, func() bool { return !proxy.State().IsAlive() })
 }
```